WIP: actors evade collision

antonkurkin opened the pull request:
Here is my proposal for agents

The idea is to try and make agents look reasonably intelligent and interract with each other using only AI lanes.
So I would want to do that that without a need for physics and colliders.

How I'm trying to do it:
- make lane keep order of agents on it  - so it could be easily searched with binary search when needed - e.g. on lane change or if only checking whether lane change is possible
- make agents keep track which agents are forward and backward of it, so it would be easibly accessible
- links between only have to be updated if an agent changes lane - or going faster than another agent and gettin on it from afar and didn't find it beforehand
- use IDM model for actor's following math so if possible an actor would stop some distance before hitting another actor - and I'm thinking that it must always be possible for not very agressive actors

I'm also thinking of a case for changing lane continuously. I'm thinking of creating temporary personal lane that would connect one lane and another when actor decides to change the lane. and creating 2 virtual agents for lanes it's moving from and to (or more if there are multiple transition lanes it intersects). this way other actors will know that somebody is getting on their lane and will be able to evade them

as for colliders - I'm thinking along the lines that all of them should have a collider that is not checked against each other and the road. then we add a collider of other group to player (and maybe other disrupting actors), and if they collide with normal actors, we change collision group and they'll be able to collide with other actors, propagating this collision group change

what doesn't work yet:
- moving backwards
- lane change and merge lanes
- workaround for intersections/broken connectivity
- smooth stopping at lane end
- in some cases its possible that an agent that changed lane and doesn't move will not be noticed
- in cases when agents do collide, the workaround is not very good
- player ignores/destroys/push away and uassign lane for collided agents (using colliders most probably)

antonkurkin committed:
actors trying to not collide

antonkurkin committed:
actors trying to not collide

TheDuckCow reviewed (commented):
I wouldn't say I have a perfect idea in my head of everything you've pulled together, but a few thoughts as request to spark some discussion:

- Firstly, I like and agree with the objective of creating a lightweight way of letting agents move around without collisions. We'll have to make some tradeoffs of course, but I think it's off to a good start
- Acknowledging that we basically will depend on smart lane greenlight/redlight management to avoid intersections at, well, intersections. But that's ok
- Seems we have to think about edge cases at points where one lane merges into another (or multiple collapsing at once into fewer lanes)
- I also have a mild feeling that we could come up with a slightly simpler solution, but that's perhaps naive on my side since I haven't gone into the weeds trying to make this change myself. Mostly pointing to the creation of `LanePosition` and `to_lane_side` and  `flip_side` (is that last one even used anywhere?)
- That being said, we could also think about an alternative as to what is being tracked. In terms of the high level approach, what do you think of this?:
  - What if instead of each actor trying to keep track of each lane and each lane's cars etc, instead each vehicle just tracked it's "next" vehicle in its motion direction? This bakes in the assumption that one car is likely only to slow down in reaction to another car, and thus generally only has one car to care about (the one it would collide with next in front of itself).
  - Then, you're just tracking the distance between you and this next vehicle (even if that goes over a new road lane transition as prior goes to next)
  - Then, the other key thing missing is the concept of the vehicle length. Could have two more vars for distance m in front of the center point and behind (something we could create some utilities to help auto calculate, or at least visualize as a sort of 3D shape in the editor one day with a gizmo to control it).
  - When it comes to changing lanes OR multiple lanes merging together and a vehicle needing to decide "if there's space", it can do the check for distance between vehicles but considering the front / back bumper distances, so that cars aren't treated as point masses. So it's kind of like a zigzag or binary tree rebalancing in a way, where a lane merge just inserts or changes the linked list of cars
  - As an example, imaging we have a LeftLane with the front bumper of CarA being 7 meters behind the rear bumper of CarB, and CarC in RightLane is looking to merge left in between CarA and CarB. If CarC has a length of 3meters, then it means there is in theory a range of 4 meters at which any point CarC could change to the left lane between CarA and CarB (ie such that CarC's rear bumper ends up still in front of CarA's front bumper, and so forth). Then when performing or initiating the merge, presuming one day we'll 'lerp' the transition more, it could trigger the routines on CarC, itself, and CarB to refresh/find what is the new "next" car and thus resetting any logic associated with that (such as trying to keep minimum distances at certain speeds etc).
  - In fact this even creates the idea of straddling lanes, maybe a car SHOULD be registered to multiple lanes at once to ensure they can be considered taking up space in both during transitions or lane changing, or even as another way to handle intersections? And then just remove the CarC in this example from the old lane once fully moved over.
  - My thought is having each agent just track at most one other vehicle (fetching it from a Road lane only when it can no longer track/find said vehicle based on registered vehicles in that lane) would simplify the number of checks and logic being done per frame, and scales just linearly with the number of vehicles
- New acceleration model: I like the idea, will need to tweak the defaults but I see the export vars still work. I do find the current model doesn't really have drag and seems to coast, I kind of liked how we had it auto slowing down, but really that's personal preferences and any dev could adjust to their own needs.

All just ideas, let me know what you think!

antonkurkin commented on addons/road-generator/nodes/road_lane_agent.gd:
yeah I'm thinking that obviously player have to have control over who to crach into and all that
I would say that in general case, player shouldn't use move along lane, and plugin's user maybe should register player on all the lanes it blocks so other cars would not crash into them - somehow I'm not planning to work on that
for my usecase I plan to have player on ai lane, which is very close to procedural generation demo, so I'm thinking of adding a path for player to ignore collisions (or remove actors they collide with)

yeah lane merging doesn't work yet. but generally speaking i plan to make it work the same way as normal lane change

assertion is because of the collision on lane the code doesn't expect multiple agents at the same offset (same point on the lane) currently. but I think I'll have to deal with that somehow later on - especially with virtual lane agents

antonkurkin commented on addons/road-generator/nodes/road_lane.gd:
```diff
@@ -87,12 +118,17 @@ var geom_node: MeshInstance3D
+# this container should contain agents in order:
+# from beginning to the end, i.e. agents' lower offset to higher offset
+# also offsets are expected but not enforced to be unique
+var _agents_in_lane: Array[RoadLaneAgent] = [] # Registration
```
yeah my thoughts exactly

antonkurkin commented on demo/procedural_generator/road_actor.gd:
```diff
@@ -126,7 +147,7 @@ func _physics_process(delta: float) -> void:
 	# Get another point a little further in front for orientation seeking,
 	# without actually moving the vehicle (ie don't update the assign lane
 	# if this margin puts us into the next lane in front)
-	var orientation:Vector3 = agent.test_move_along_lane(0.05)
+	var orientation:Vector3 = agent.test_move_along_lane(rotate_to_distance)
 
 	if ! global_transform.origin.is_equal_approx(orientation):
 		look_at(orientation, Vector3.UP)
```
oh nice. will do that

antonkurkin replied:
on the other hand, I think it's less of a problem with agent remembering offset move along lane shouldn't be too heavy
and results should be marginally better it we calculate angle from the positions of front and back wheels on lane

I also thinking that it may be worth it to calculate distance to agent on lane instead of using distance_to

antonkurkin commented on demo/procedural_generator/road_actor.gd:
```diff
@@ -54,13 +57,28 @@ func get_input() -> Vector3:
 
+## For more info see Intelligent driver model
+## https://en.wikipedia.org/wiki/Intelligent_driver_model
+func compute_idm_acceleration(next_agent: RoadLaneAgent) -> float:
```
no the idea as far as I understood - delta in this case is constant for kind of emulating acceleration/deceleration profile
it's recommended to be 4 but we could play around with it (and other actor's characteristics) to make it look more realistic

antonkurkin commented on demo/procedural_generator/procedural_generator.gd:
```diff
@@ -137,35 +141,35 @@ func add_next_rp(rp: RoadPoint, dir: int) -> void:
-	for _lane in new_lanes:
-		var rand_offset = randf() * _lane.curve.get_baked_length()
+	for _lane: RoadLane in new_lanes:
+		var rand_offset = randf() * (_lane.curve.get_baked_length() - before_end)
```
agreed, more of a workaround to not create colliding agents for now

antonkurkin commented on addons/road-generator/nodes/road_lane.gd:
```diff
@@ -138,28 +174,91 @@ func get_lane_end() -> Vector3:
-func register_vehicle(vehicle: Node) -> void:
-	_vehicles_in_lane.append(vehicle)
+func is_agent_list_correct():
+	return true
```
:) just forgot to enable it back after playing around without a sanity check, good catch

antonkurkin commented on addons/road-generator/nodes/road_lane.gd:
```diff
@@ -16,20 +16,47 @@ const COLOR_START := Color(0.7, 0.7, 0,7)
+	set(val): side_lanes[LaneSideways.RIGHT] = val
+
+
+var adjacent_lanes: Array[NodePath] = ["", ""]
```
yeah I'm not too happy with namings I have either:)

antonkurkin commented on addons/road-generator/nodes/road_lane.gd:
```diff
@@ -138,28 +174,91 @@ func get_lane_end() -> Vector3:
+func register_agent(agent: RoadLaneAgent, link_agents: bool) -> void:
+	if DEBUG_OUT:
+		print(Time.get_ticks_usec(), " Registering agent ", agent, " on lane ", self)
+	assert(self.is_agent_list_correct())
+	assert(agent not in _agents_in_lane)
+	assert(agent.lane_pos)
```
on one hand I wouldn't want to have heavy checks like is_agent_list_correct in production. on the other, I suppose debugging players' problems with error logs (that I don't really do much normally) need to be more of a priority for me. some of the asserts should be errors, that's for sure

TheDuckCow commented:
Some more thoughts for ya, good discussion points indeed:

> but with more traffic I'm thinking of maybe some managing code to decide who goes next

Though there's always a risk of deadlocking, I tend to like the options of letting individual agents decide for themselves what to do. As a hypothetical more sophisticated example of this, in an intersection, we can provide the input that a lane is "disabled" (red light), but it's up to the agent to actually stop (or maybe they eased into the intersection for a left turn, they could decide to still go for it once oncoming traffic is cleared).

So on that basis, I'm more in favor of having the vehicles decide amongst themselves how to merge. That logic for this level of behavior should definitely be in the road actor/procedural scene, since there's no single correct way to determine this, so our implementation there would just be an example.

> so I'm thinking of getting rid of it - and just adding offset field to the agent

Yeah I feel that could simplify things with the same end benefit. And if we can essentially cache an agents offset from a prior test/move, then it saves from processing later. In fact, that makes the physical space calculation rather easy because offset is already a figure in meters. the distance between two agents is `(agent_a.lane_offset + agent_a.front_offset)` (ie the bumper) `- (agent_b.lane_offset - agent_b.rear_offset)` without ever having to do more curve math, presuming both vehicle's current offset is reasonably up to date.

> Yeah I think I'll do that and forbid non-player actors to move backward for now (as it's kind of an edge case), I don't use backward link much anyway

I'm fine with such a restriction for now if it makes things easier/simpler, this code is just demo anyways so it's ok for us to call out restrictions to make the demo as simple as possible.

> Do we want to have it in agent or in actor? it kind of make sense to have prolonged agent to analyze ai lanes' usage and it would definitely simplify things for agent logic. On the same note I was actually thinking of moving agent's position to the front of an actor. this way we only need to store one size variable and it would also stop automatically at the lane's end for example (but moving backward it won't be automatic) though I suppose rotation of the vehicle would be a bit worse if we use position on the front to pick for direction

I'd suggest the property to be on the agent as it would indeed be useful for other purposes

I also mentally contemplated putting the offset at one side or the other, but then you still need the size input for how far backwards to read the next vehicle, and turning the agent into a spatial feels... well overkill is a strong word, but I worry even in my own games about the number of nodes as each one has overhead. As just reference as it is now, it remains a bit lighter weight and can rely on the transform of the 3D nodes that inevitably already exist.

> we may need some tricks to figure out when to create and when to remove them but I don't see it being too much of a problem.

I see it as more like a "promise" that the agent holds to release its virtual registration at some point, that way nobody needs to take care of it other than the vehicle itself. I'm not as in favor of creating new agents on the fly or duplicate paths or anything like that because it becomes quite a bit of overhead (every new node added to the scene adds up), but that does raise the question of how we keep track of the "virtual" registrations and their offsets. Could be another array of virtual registrations which could be an array of [lane, offset] pairs. Or, it could be your lane offset class you created already I suppose (though that becomes some class memory overhead, not as much as a node however). 

> I'm also thinking of a case for changing lane continuously. I'm thinking of creating temporary personal lane that would connect one lane and another when actor decides to change the lane. and creating 2 virtual agents for lanes it's moving from and to (or more if there are multiple transition lanes it intersects).

(my above point also address this part from your original post):

This level of behavior nuance also starts to get into the weeds of traffic simulation. I want to avoid forcing the RoadLaneAgent into a one size fits all solution. It's not meant to be "traffic working out of the box", but rather just provide the tools to make it easy to implement traffic. Then, the _demo_ scenes are where we provide potentially different example implementations. At some point, I want to have a carbody physics based example to contrast our "on rails" solution.

So, I'm a bit torn how much of this specific merging like logic should be in the road lane agent (which is provided to all road generator users whether they like it or not, and thus making it more complicated to understand), or trying to do more of this logic in the road actor and potentially keeping track of additional lanes there. Another idea could be that the road actor scene actually creates a new inherited version of the road lane agent that adds additional variables, showing people they have their own way of extending it further if they want. Just ideas, depending on where we go.

In terms of what should go into the RoadLaneActor vs not, I'd say: utilities or properties that would be universal for any traffic system are fine, but workflow specific ones should be pushed into the demo files. Length of a vehicle is a pretty universal thing to keep track of, but the exact way and who decides to go during a zipper merge is better for the demo code (since there are many strategies to avoid deadlocking where traffic just gets jammed or one lane never gets to go).

> yeah I also want player's auto slowing down back (and maybe heavier breaking on pressing down). will look at how it can be done

👍 cool, but super low priority compared to the rest which is the more functional parts.

> as for colliders - I'm thinking along the lines that all of them should have a collider that is not checked against each other and the road. then we add a collider of other group to player (and maybe other disrupting actors), and if they collide with normal actors, we change collision group and they'll be able to collide with other actors, propagating this collision group change

Since I didn't respond to this part before, I think it could be interesting to have a pure no- collider demo example, knowing that I intend at some point to have collider/carbody based counter demo at some point too. I would suggest we try to keep each individual demo as simple as is reasonably possible, so it makes it easier for someone looking at the code better understand how it all fits together. One day we could always have a third "super demo" which is something like combining both concepts together (cars on rails in the distance, but simulated up close to the camera with collisions)

antonkurkin commented:
> I also mentally contemplated putting the offset at one side or the other, but then you still need the size input for how far backwards to read the next vehicle, and turning the agent into a spatial feels... well overkill is a strong word, but I worry even in my own games about the number of nodes as each one has overhead. As just reference as it is now, it remains a bit lighter weight and can rely on the transform of the 3D nodes that inevitably already exist.
I think we could ask users to just place actor geometry with a transformation that makes beginning/front wheels the position for  actor (and so agent). but two additional offsets are fine as well, I think it gives some flexibility without much complexity/resource demand from us

> I see it as more like a "promise" that the agent holds to release its virtual registration at some point, that way nobody needs to take care of it other than the vehicle itself. I'm not as in favor of creating new agents on the fly or duplicate paths or anything like that because it becomes quite a bit of overhead (every new node added to the scene adds up), but that does raise the question of how we keep track of the "virtual" registrations and their offsets. Could be another array of virtual registrations which could be an array of [lane, offset] pairs. Or, it could be your lane offset class you created already I suppose (though that becomes some class memory overhead, not as much as a node however).
oh yeah agents should track/remove their virtual agents when it doesn't need them anymore, that's for sure.
By tricks I meant some kind of logic in the agent that would decide when to put agent 
I have to say that think for this PR I indend to just use RoadLaneActor clones.
But we may also want to separate kind of RoadLaneObstacle parent for RoadLaneAgent (with current_lane, lane offset, front/back offsets and speed) that we could reuse for basically everything that blocks a part of RoadLane

> In terms of what should go into the RoadLaneActor vs not, I'd say: utilities or properties that would be universal for any traffic system are fine, but workflow specific ones should be pushed into the demo files. Length of a vehicle is a pretty universal thing to keep track of, but the exact way and who decides to go during a zipper merge is better for the demo code (since there are many strategies to avoid deadlocking where traffic just gets jammed or one lane never gets to go).
ok that's fair, I'm fine either way. I'd argue that some "dead" (like not used in a specific project) code is not too bad, if it is marked as used for specific stuff and could be "widely" applicable to other projects.
e.g. if your actors are fine moving without rails and just use road agent as kind of a reference point to where they are trying to go, so your actor doesn't call some functionality of RoadLaneAgent. but anyway I think we could discuss the split of what goes in the demo and what goes in plugin code when it is closer to completion and we know how it looks. I'll keep it in mind

> Since I didn't respond to this part before, I think it could be interesting to have a pure no- collider demo example, knowing that I intend at some point to have collider/carbody based counter demo at some point too. I would suggest we try to keep each individual demo as simple as is reasonably possible, so it makes it easier for someone looking at the code better understand how it all fits together. One day we could always have a third "super demo" which is something like combining both concepts together (cars on rails in the distance, but simulated up close to the camera with collisions)
yeah I don't think of adding anything physics related in this PR, just thoughts. sounds good.

could you also look at how you want to achieve intersection/multi-container lane linkage? I don't mind trying to implement that when I get to it - ideally in this PR, as without it everything's going to kind of break on this transition - but I didn't look at the idea of how it should be done

antonkurkin committed:
connect RoadLanes when roadpoint is connected and a segment is created

antonkurkin committed:
connect RoadLanes when roadpoint is connected and a segment is created

antonkurkin commented:
@TheDuckCow ah I remembered why I wanted to have both forward and backward links - in case if traffic comes to stop, we could put actors to sleep and when the actor in front of them starts moving again, they would wake them up

antonkurkin committed:
changes according to PR discussions

started preparation for agent blocking multiple lanes
separated road lane obstacle from road lane agent
separated move_along_lane to an object - may need to use more than one

removed agent links to next/prior
added end offsets to obstacles

antonkurkin committed:
changes according to PR discussions

started preparation for agent blocking multiple lanes
separated road lane obstacle from road lane agent
separated move_along_lane to an object - may need to use more than one

removed agent links to next/prior
added end offsets to obstacles

antonkurkin commented:
changed the idea, now tracking of obstacles is only done in lanes
I didn't yet look at other demos and there are still some bugs in procedural demo that I have to track down (sometimes actors just start to move backward with insane speed which is pretty fun to look at)
but you can still try it and see what's the current progress and how I'm planning to achieve the goal of this PR

antonkurkin committed:
make actors react to collision and lane end

antonkurkin committed:
make actors react to collision and lane end

antonkurkin commented:
the problem was about reusing the actors (not cleaning up the speed) and it looks like IDM has some problem where it can get to negative speed and just unable to get out of it. updated it now with just speed clamping
also made actors to push each other. actually pretty fun to play with
still didn't do proper lane merging but now allow actors to do that just for the fun of it so some asserts may be triggered in rare cases

antonkurkin committed:
add lane description (merging/diverging/etc) for decision making purposes

antonkurkin committed:
add lane description (merging/diverging/etc) for decision making purposes

antonkurkin committed:
implemented shared part of the lane

antonkurkin committed:
implemented shared part of the lane

antonkurkin committed:
fix spawner and make IDM behave reasonable when moving in reverse

antonkurkin committed:
fix spawner and make IDM behave reasonable when moving in reverse

antonkurkin committed:
fix most of the asserts

antonkurkin committed:
fix most of the asserts

antonkurkin committed:
start using obstacle looked ahead to decide if there is a collision

antonkurkin committed:
start using obstacle looked ahead to decide if there is a collision

antonkurkin committed:
fix switching lanes on shared parts

antonkurkin committed:
fix switching lanes on shared parts

antonkurkin committed:
some uncommitted work on previous idea

antonkurkin committed:
some uncommitted work on previous idea

antonkurkin committed:
fixups after rebase

antonkurkin committed:
fixups after rebase

antonkurkin committed:
remove all the shared part related stuff

antonkurkin committed:
remove all the shared part related stuff

antonkurkin commented:
removed my implementation of shared part and everything related to obstacle search excluding some stubs where I plan to try to implement the proposed traffic search/decision making in the next commit

antonkurkin committed:
add obstacle list

antonkurkin committed:
add obstacle list

TheDuckCow commented on addons/road-generator/nodes/road_lane.gd:
```diff
@@ -94,15 +197,41 @@ var this_road_segment = null # RoadSegment
+	MERGING = 0x2, # secondary lane that merges into the primary lane (mutually exclusive with MERGE_INTO, see merge_lane)
+	DIVERGE_FROM = 0x4, # main lane from which secondary lanes diverging from (mutually exclusive with DIVERGING)
+	DIVERGING = 0x8, # secondary lane that diverges from the primary lane (mutually exclusive with DIVERGE_FROM, see diverge_from)
+	INTERSECTION = 0x10, # the lane is a part of intersection. it may intersect other lanes (see intersection_points)
```
Is there special meaning to these particular values? up till this one, looked like it was following bit flag logic (but I do get the point that the values one comparing to the other is how they compare by weight)

antonkurkin replied:
these are bit flags - 2^n (0x10 hexadecimal is 16 decimal, 0x20 is 32). or am I missing something? I just think that it's easier to read and not make mistakes for bigger flags:)

TheDuckCow commented on demo/procedural_generator/road_actor_spawner.gd:
```diff
@@ -2,18 +2,26 @@ extends Node3D
 
 ## simple RoadLane with override to despawn anyone assigned to it
 class DespawnRoadLane extends RoadLane:
+	#TODO should be @tool?
```
I'm fine if you make this a tool, especially being in the demo folders

TheDuckCow reviewed (commented):
Left a variety of comments and thoughts, as mentioned might be good to have a followup sync at some point, but hopefully this gives you enough direction for now.

antonkurkin commented on addons/road-generator/nodes/road_lane.gd:
sounds reasonable to me. I think if RoadLaneAgent is not used (and so obstacles are not used) then the only overhead in the new approach is going to be the search array for next obstacle on lane. again, that, I think can be moved to trafficmanager (as a map from road lane to array) but then we need to keep track of lane changes (geometry or sequiential links) there as well

another approach could be - in the stuff that I do now (not yet committed), I use a constant for the length of the chunk (and so how many pointers the search array has), but I think I can move it to RoadManager and if length is 0 then don't initialize the array, and the only overhead this way is going to be an empty array in RoadLane

TheDuckCow commented on addons/road-generator/nodes/road_lane.gd:
```diff
@@ -125,9 +254,9 @@ func _ready():
-			if is_instance_valid(_vehicle):
-				_vehicle.call_deferred("queue_free")
+		for obstable in obstacles:
+			if is_instance_valid(obstable):
```
I think obstacle is a fine general name. While "vehicle" made be a bit more obvious to some people if, it still makes sense we end up calling it an obstacle. One question though - if we wanted to make a stationary object, do we still always have to attach a roadlaneagent? Would we need some kind of container called roadlanestatic or something?

No worries on functionality, we can do fullter tester when it's further along - the enum option was just an idea

antonkurkin replied:
in my head user don't always need a RoadLaneAgent, instead they can just place an Obstacle object on lane (though they'll have to do it themself) - I think that should be kind of the normal interface object to seek for, well, any obstacles on the road

TheDuckCow commented on addons/road-generator/nodes/road_lane.gd:
```diff
@@ -15,31 +15,120 @@ extends Path3D
+class Obstacle:
+	enum ObstacleFlags {
+		REAL = 0x0, # the node is on this lane
+		IMMINENT = 0x1, # the node from anothe lane won't be able to stop before it gets to this position
```
Yeah seems reasonable to me that within any given traffic system, you could make it work. In real life, it's ultimately always going to be the judgement of the driver whether the _other_ driver is going to yield or not (regardless of laws) and so seems reasonable that INTENT, PARTIAL, and REAL would cover the actual feasible states.

If though it really complicates your approach, then we could still have the IMMINENT option, so long as it's ultimately the road actors that define the state and not the agent code itself, since it's highly dependent on how the traffic solution is working.

antonkurkin replied:
I don't need it right now, probably will be necessary when we get to traffic on intersections, but I'll leave it be for now.

antonkurkin commented on addons/road-generator/nodes/road_lane.gd:
```diff
@@ -15,31 +15,120 @@ extends Path3D
+	set(val): assert(get_node_or_null(val) != self); side_lanes[SideDir.RIGHT] = val
+
+
+var sequential_lanes: Array[NodePath] = ["", ""]
```
not sure I understand, do you mean i should initialize it with `null` or something else?

currently it's the same way as we had it before - so the RoadLane has next_lane and prior_lane, they are always there. but I changed the way they are implemented with an array so we could access them by index. but to be able to access prior/next lane (with actual value or null) we need it to be initialized with 2 elements.

if we start without elements then we ned to add elements as we go and if we want only prior (index 1) link than we need to initialize next (index 0) as well

antonkurkin committed:
use multiple despawn lanes, so lane connections/end obstacles would be properly linked

antonkurkin committed:
use multiple despawn lanes, so lane connections/end obstacles would be properly linked

antonkurkin committed:
update ActorSpawner for better logging

antonkurkin committed:
update ActorSpawner for better logging

antonkurkin committed:
fix some errors

antonkurkin committed:
fix some errors

antonkurkin committed:
fix after rebase

antonkurkin committed:
fix after rebase

antonkurkin committed:
fix obstacle links on lane change, improve obstacle move_along_lane

antonkurkin committed:
fixes after rebase

antonkurkin committed:
implelment more robust lane disconnection to fix some problems

antonkurkin committed:
update despawn lanes geometry to continue from the lane it linked to

antonkurkin committed:
make actors not stopping before an end of the lane obstacle

antonkurkin committed:
fix a problem with connection/disconnection breaking _next_obstacles, make despawn deferred to fix obstacle list, add more asserts for debugging

antonkurkin committed:
fix obstacles being broken during moving obstacle along lane

antonkurkin committed:
fix end obstacle on lane change

antonkurkin committed:
fix initialization of end obstacles

antonkurkin committed:
generate _next_obstacles on lanes from editor

antonkurkin committed:
implement 3-tier distance calculation

antonkurkin committed:
update road actor to drag rear axle behind front axle

antonkurkin committed:
move obstacle to a separate file. do some renaming

antonkurkin committed:
updated to disable the obstacle search and linkage

antonkurkin committed:
added function descriptions, fixed some problems with disabling the obstacle lists

antonkurkin committed:
fix memory leaks and improper road lane connection/disconnection when using inspector or removing from tree. cleaned up some things and added some TODOs

antonkurkin committed:
added comments. fixd some issues from the previous cleanup...

antonkurkin committed:
fix incorrect lane order for multiple obstacles in the same _next_bstacle chunk

code deduplication
more cleanups

antonkurkin committed:
prepare obstacle list functionality for backward motion. actor code is in progress

antonkurkin committed:
update road actor to work with obstacles while moving backwards

antonkurkin committed:
initial implementation for road actor to actually see obstacles on merging/diverging lanes. fixed how primary lanes are set

antonkurkin committed:
improve merging/diverging

antonkurkin committed:
cleanup move along lane

I think it was changed like that to have quite a bit more functionality (like stopping on obstacle)

antonkurkin committed:
enable disconnected lane hopping workaround. various cleanups in road lane/agent/obstacle

antonkurkin commented on addons/road-generator/nodes/road_lane.gd:
```diff
@@ -422,7 +418,11 @@ func disconnect_side(dir :SideDir) -> void:
-	#TODO disconnect all the merging/diverging from primary if necessary. assign new primary if previous one is disconnected? change flags?
+	for lane:RoadLane in [self, lane_side]:
+		if (lane.flags & (RoadLane.Flags.DIVERGING | RoadLane.Flags.MERGING)) != 0:
+			push_warning("disconnecting ", lane , " that is marked as merging/diverging. its primary lanes are ", lane._primary_lanes[MoveDir.FORWARD], " and ", lane._primary_lanes[MoveDir.BACKWARD])
```
@TheDuckCow not sure what to do about this one. any ideas?

antonkurkin commented on addons/road-generator/nodes/road_lane.gd:
```diff
@@ -250,16 +250,12 @@ func _exit_tree() -> void:
 		self.disconnect_sequential(dir)
 	for dir in SideDir.values():
 		self.disconnect_side(dir)
-	assert(auto_free_vehicles != (self.traffic_chunk_length > 0)) #TODO support something like road_actor_manager for despawn.
```
@TheDuckCow what are your thoughts on supporting actor pooling in RoadLane for automatic despawn? I think we could have some kind of a pure virtual interface for it - or to just have it as Node and check if it has a `remove_actor` method to call. and if assigned in roadmanager use it instead of queue_free

antonkurkin committed:
some cleanups

antonkurkin committed:
implement simple traffic light

changed obstacle flags to type

antonkurkin committed:
copy RoadActor to SimpleRoadActor to preserve history after all the obstacle related changes

antonkurkin committed:
replace RoadActor with simple road actor, move obstacle aware one to separate scene and script

antonkurkin committed:
update script link in the obstacle aware road actor scene, update links in demo scenes

antonkurkin committed:
cleanup after simple road actor rollback

setup uids (also for traffic light)
update simple road actor to work with current road agent
update obstacle sanity checks in case of no line registration

antonkurkin committed:
small fixes in traffic light initialization

antonkurkin committed:
refactor road actor to search obstacles more consistently, work around for distance calculation to traffic light stop light on side lane
